File: java/src/test/java/com/warehouse/application/PlaceOrderUseCaseTest.java
package com.warehouse.application;

import static org.assertj.core.api.Assertions.assertThat;

import com.warehouse.adapters.FakePaymentProcessor;
import com.warehouse.adapters.InMemoryInventoryGateway;
import com.warehouse.adapters.InMemoryOrderRepository;
import com.warehouse.application.InventoryGateway.ReleaseOutcome;
import com.warehouse.application.InventoryGateway.ReservationOutcome;
import com.warehouse.application.InventoryGateway.ReservationToken;
import com.warehouse.application.OrderRepository.SaveOutcome;
import com.warehouse.application.PaymentProcessor.ChargeOutcome;
import com.warehouse.domain.DomainError.CompensationFailureError;
import com.warehouse.domain.DomainError.InsufficientStockError;
import com.warehouse.domain.DomainError.InvalidOrderError;
import com.warehouse.domain.DomainError.PaymentDeclinedError;
import com.warehouse.domain.DomainError.PersistenceConflictError;
import com.warehouse.domain.Money;
import com.warehouse.domain.Order;
import com.warehouse.domain.OrderId;
import com.warehouse.domain.OrderLine;
import com.warehouse.domain.Quantity;
import com.warehouse.domain.Sku;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;

/**
 * Integration tests: the full place-order pipeline over in-memory adapters, covering the happy path
 * plus each failure path (CONTRACTS.md §2).
 */
class PlaceOrderUseCaseTest {

  private static final String USD = "USD";

  private static OrderLine line(String sku, int quantity, long minorUnits) {
    return new OrderLine(new Sku(sku), new Quantity(quantity), new Money(minorUnits, USD));
  }

  /** The use case plus the doubles it was wired to, for interaction asserts. */
  private record Pipeline(
      PlaceOrderUseCase useCase,
      InMemoryInventoryGateway inventory,
      FakePaymentProcessor payments,
      InMemoryOrderRepository repository) {

    static Pipeline create(Map<String, Integer> stock, boolean decline) {
      Map<Sku, Integer> initialStock =
          stock.entrySet().stream()
              .collect(
                  java.util.stream.Collectors.toMap(e -> new Sku(e.getKey()), Map.Entry::getValue));
      var inventory = new InMemoryInventoryGateway(initialStock);
      var payments = new FakePaymentProcessor(decline);
      var repository = new InMemoryOrderRepository();
      var seq = new AtomicInteger();
      OrderIdGenerator ids = () -> new OrderId("ord-" + seq.incrementAndGet());
      return new Pipeline(
          new PlaceOrderUseCase(inventory, payments, repository, ids),
          inventory,
          payments,
          repository);
    }
  }

  @Test
  void happyPathReservesChargesPaysAndPersistsPaid() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), false);

    var result = pipeline.useCase().execute(List.of(line("SKU-1", 2, 500)), "idem-1");

    assertThat(result)
        .isInstanceOfSatisfying(
            PlaceOrderResult.Success.class,
            success -> {
              assertThat(success.order().state()).isEqualTo(Order.State.PAID);
              assertThat(success.order().id().value()).isEqualTo("ord-1");
              assertThat(success.order().total()).isEqualTo(new Money(1000, USD));
              assertThat(success.order().version()).isEqualTo(1);
              assertThat(pipeline.repository().get(success.order().id()))
                  .isPresent()
                  .hasValueSatisfying(
                      stored -> {
                        assertThat(stored.state()).isEqualTo(Order.State.PAID);
                        assertThat(stored).isNotSameAs(success.order());
                      });
            });
    assertThat(pipeline.payments().chargedOrders()).hasSize(1);
    assertThat(pipeline.repository().saved()).hasSize(1);
    assertThat(pipeline.inventory().snapshotStock().get(new Sku("SKU-1"))).isEqualTo(8);
  }

  @Test
  void insufficientStockFailsWithoutChargeOrPersist() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 1), false);

    var result = pipeline.useCase().execute(List.of(line("SKU-1", 5, 500)), "idem-2");

    assertThat(result)
        .isInstanceOfSatisfying(
            PlaceOrderResult.Failure.class,
            failure -> {
              assertThat(failure.error()).isInstanceOf(InsufficientStockError.class);
              assertThat(((InsufficientStockError) failure.error()).available()).isEqualTo(1);
            });
    assertThat(pipeline.payments().chargedOrders()).isEmpty();
    assertThat(pipeline.repository().saved()).isEmpty();
    assertThat(pipeline.inventory().snapshotStock().get(new Sku("SKU-1"))).isEqualTo(1);
  }

  @Test
  void paymentDeclineReleasesReservationWithoutPersist() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), true);

    var result = pipeline.useCase().execute(List.of(line("SKU-1", 1, 500)), "idem-3");

    assertThat(result)
        .isInstanceOfSatisfying(
            PlaceOrderResult.Failure.class,
            failure -> {
              assertThat(failure.error()).isInstanceOf(PaymentDeclinedError.class);
              assertThat(((PaymentDeclinedError) failure.error()).reason()).contains("declined");
            });
    assertThat(pipeline.payments().chargedOrders()).hasSize(1);
    assertThat(pipeline.repository().saved()).isEmpty();
    assertThat(pipeline.inventory().snapshotStock().get(new Sku("SKU-1"))).isEqualTo(10);
  }

  @Test
  void saveFailureRefundsAndReleases() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), false);
    pipeline.repository().failSave(true);

    var result = pipeline.useCase().execute(List.of(line("SKU-1", 1, 500)), "idem-4");

    assertThat(result)
        .isInstanceOfSatisfying(
            PlaceOrderResult.Failure.class,
            failure -> assertThat(failure.error()).isInstanceOf(PersistenceConflictError.class));
    assertThat(pipeline.payments().refunded()).hasSize(1);
    assertThat(pipeline.inventory().snapshotStock().get(new Sku("SKU-1"))).isEqualTo(10);
    assertThat(pipeline.repository().saved()).isEmpty();
  }

  @Test
  void compensationFailureAfterSaveFailure() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), false);
    pipeline.repository().failSave(true);
    pipeline.payments().failRefund(true);

    var result = pipeline.useCase().execute(List.of(line("SKU-1", 1, 500)), "idem-5");

    assertThat(result)
        .isInstanceOfSatisfying(
            PlaceOrderResult.Failure.class,
            failure -> assertThat(failure.error()).isInstanceOf(CompensationFailureError.class));
  }

  @Test
  void releaseFailureAfterSaveFailureIsCompensationFailure() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), false);
    pipeline.repository().failSave(true);
    pipeline.inventory().failRelease(true);

    var result = pipeline.useCase().execute(List.of(line("SKU-1", 1, 500)), "idem-5c");

    assertThat(result)
        .isInstanceOfSatisfying(
            PlaceOrderResult.Failure.class,
            failure -> assertThat(failure.error()).isInstanceOf(CompensationFailureError.class));
    assertThat(pipeline.payments().refunded()).hasSize(1);
  }

  @Test
  void releaseFailureAfterDeclineIsCompensationFailure() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), true);
    pipeline.inventory().failRelease(true);

    var result = pipeline.useCase().execute(List.of(line("SKU-1", 1, 500)), "idem-5b");

    assertThat(result)
        .isInstanceOfSatisfying(
            PlaceOrderResult.Failure.class,
            failure -> assertThat(failure.error()).isInstanceOf(CompensationFailureError.class));
  }

  @Test
  void invalidLinesFailValidationWithoutTouchingPorts() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), false);

    var result = pipeline.useCase().execute(List.of(), "idem-6");

    assertThat(result)
        .isInstanceOfSatisfying(
            PlaceOrderResult.Failure.class,
            failure -> assertThat(failure.error()).isInstanceOf(InvalidOrderError.class));
    assertThat(pipeline.payments().chargedOrders()).isEmpty();
    assertThat(pipeline.repository().saved()).isEmpty();
    assertThat(pipeline.inventory().snapshotStock().get(new Sku("SKU-1"))).isEqualTo(10);
  }

  @Test
  void getReturnsEmptyForUnknownIdentifier() {
    var repository = new InMemoryOrderRepository();
    assertThat(repository.get(new OrderId("missing-order"))).isEmpty();
  }

  @Test
  void reserveAllIsIdempotentForTheSameKey() {
    var inventory = new InMemoryInventoryGateway(Map.of(new Sku("SKU-1"), 2));
    var lines = List.of(line("SKU-1", 1, 100));
    var first = inventory.reserveAll(new OrderId("ord-1"), lines, "idem-x");
    var second = inventory.reserveAll(new OrderId("ord-1"), lines, "idem-x");
    assertThat(first).isInstanceOf(ReservationOutcome.Reserved.class);
    assertThat(second).isInstanceOf(ReservationOutcome.Reserved.class);
    assertThat(inventory.snapshotStock().get(new Sku("SKU-1"))).isEqualTo(1);
  }

  @Test
  void releaseOfUnknownTokenSucceeds() {
    var inventory = new InMemoryInventoryGateway(Map.of());
    var outcome = inventory.release(new ReservationToken(new OrderId("ord-1"), "missing"));
    assertThat(outcome).isInstanceOf(ReleaseOutcome.Released.class);
  }

  @Test
  void chargeIsIdempotentForTheSameKey() {
    var payments = new FakePaymentProcessor(false);
    var order = new Order(new OrderId("ord-1"), List.of(line("SKU-1", 1, 100)));
    var first = payments.charge(order, "idem-pay");
    var second = payments.charge(order, "idem-pay");
    assertThat(first).isInstanceOf(ChargeOutcome.Charged.class);
    assertThat(second).isInstanceOf(ChargeOutcome.Charged.class);
    assertThat(payments.chargedOrders()).hasSize(1);
  }

  @Test
  void saveThenGetReturnsDetachedPaidSnapshot() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), false);
    var result = pipeline.useCase().execute(List.of(line("SKU-1", 1, 100)), "idem-repo");
    var success = (PlaceOrderResult.Success) result;

    var loaded = pipeline.repository().get(success.order().id()).orElseThrow();
    loaded.ship();
    var stored = pipeline.repository().get(success.order().id()).orElseThrow();
    assertThat(stored.state()).isEqualTo(Order.State.PAID);
    assertThat(stored.version()).isEqualTo(1);
  }

  @Test
  void saveWithStaleVersionIsPersistenceConflict() {
    var repository = new InMemoryOrderRepository();
    var order = new Order(new OrderId("ord-2"), List.of(line("SKU-1", 1, 100)));
    order.pay();
    assertThat(repository.save(order, 0)).isInstanceOf(OrderRepository.SaveOutcome.Saved.class);
    var retry = order.snapshot();
    retry.ship();
    assertThat(repository.save(retry, 0)).isInstanceOf(SaveOutcome.Conflict.class);
    assertThat(repository.save(order, 1)).isInstanceOf(SaveOutcome.Saved.class);
  }

  @Test
  void idempotentReplayDoesNotDoubleCharge() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), false);
    var first = pipeline.useCase().execute(List.of(line("SKU-1", 2, 300)), "idem-7");
    var second = pipeline.useCase().execute(List.of(line("SKU-1", 2, 300)), "idem-7");

    assertThat(first).isInstanceOf(PlaceOrderResult.Success.class);
    assertThat(second).isInstanceOf(PlaceOrderResult.Success.class);
    assertThat(((PlaceOrderResult.Success) second).order().state()).isEqualTo(Order.State.PAID);
    assertThat(pipeline.payments().chargedOrders()).hasSize(1);
    assertThat(pipeline.inventory().snapshotStock().get(new Sku("SKU-1"))).isEqualTo(8);
  }

  @Test
  void reusedKeyWithDifferentPayloadIsInvalidOrder() {
    var pipeline = Pipeline.create(Map.of("SKU-1", 10), false);
    pipeline.useCase().execute(List.of(line("SKU-1", 1, 100)), "idem-8");

    var result = pipeline.useCase().execute(List.of(line("SKU-1", 2, 100)), "idem-8");

    assertThat(result)
        .isInstanceOfSatisfying(
            PlaceOrderResult.Failure.class,
            failure -> {
              assertThat(failure.error()).isInstanceOf(InvalidOrderError.class);
              assertThat(((InvalidOrderError) failure.error()).reason()).contains("idempotency");
            });
    assertThat(pipeline.payments().chargedOrders()).hasSize(1);
  }

  @Test
  void concurrentReservationsDoNotOversell() throws Exception {
    var pipeline = Pipeline.create(Map.of("SKU-A", 5), false);
    List<OrderLine> lines = List.of(line("SKU-A", 5, 100));
    List<Callable<PlaceOrderResult>> tasks =
        List.of(
            () -> pipeline.useCase().execute(lines, "idem-9a"),
            () -> pipeline.useCase().execute(lines, "idem-9b"));

    try (var executor = Executors.newFixedThreadPool(2)) {
      var futures = executor.invokeAll(tasks);
      var results = futures.stream().map(PlaceOrderUseCaseTest::join).toList();
      long successes = results.stream().filter(PlaceOrderResult.Success.class::isInstance).count();
      long shortages =
          results.stream()
              .filter(PlaceOrderResult.Failure.class::isInstance)
              .map(PlaceOrderResult.Failure.class::cast)
              .filter(failure -> failure.error() instanceof InsufficientStockError)
              .count();
      assertThat(successes).isEqualTo(1);
      assertThat(shortages).isEqualTo(1);
    }
    assertThat(pipeline.inventory().snapshotStock().get(new Sku("SKU-A"))).isZero();
    assertThat(pipeline.payments().chargedOrders()).hasSize(1);
  }

  private static PlaceOrderResult join(java.util.concurrent.Future<PlaceOrderResult> future) {
    try {
      return future.get();
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new AssertionError(e);
    } catch (java.util.concurrent.ExecutionException e) {
      throw new AssertionError(e);
    }
  }
}

File: kotlin/src/test/kotlin/com/warehouse/application/PlaceOrderUseCaseTest.kt
package com.warehouse.application

import com.warehouse.adapters.FakePaymentProcessor
import com.warehouse.adapters.InMemoryInventoryGateway
import com.warehouse.adapters.InMemoryOrderRepository
import com.warehouse.application.InventoryGateway.ReleaseOutcome
import com.warehouse.application.InventoryGateway.ReservationOutcome
import com.warehouse.application.InventoryGateway.ReservationToken
import com.warehouse.application.OrderRepository.SaveOutcome
import com.warehouse.application.PaymentProcessor.ChargeOutcome
import com.warehouse.domain.CompensationFailureError
import com.warehouse.domain.InsufficientStockError
import com.warehouse.domain.InvalidOrderError
import com.warehouse.domain.Money
import com.warehouse.domain.Order
import com.warehouse.domain.OrderId
import com.warehouse.domain.OrderLine
import com.warehouse.domain.PaymentDeclinedError
import com.warehouse.domain.PersistenceConflictError
import com.warehouse.domain.Quantity
import com.warehouse.domain.Sku
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertInstanceOf
import org.junit.jupiter.api.Assertions.assertNotSame
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.atomic.AtomicInteger

/**
 * Integration tests: the full place-order pipeline over in-memory adapters,
 * covering the happy path plus each failure path (CONTRACTS.md §2).
 */
class PlaceOrderUseCaseTest {
    @Test
    fun happyPathReservesChargesPaysAndPersistsPaid() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = false)

        val result = pipeline.useCase.execute(listOf(line("SKU-1", 2, 500)), "idem-1")

        val success = assertInstanceOf(PlaceOrderResult.Success::class.java, result)
        assertEquals(Order.Status.PAID, success.order.status)
        assertEquals("ord-1", success.order.id.value)
        assertEquals(Money(1000, USD), success.order.total())
        assertEquals(1, success.order.version)
        val stored = pipeline.repository.get(success.order.id)
        assertEquals(Order.Status.PAID, stored!!.status)
        assertNotSame(success.order, stored)
        assertEquals(1, pipeline.payments.chargedOrders().size)
        assertEquals(1, pipeline.repository.saved().size)
        assertEquals(8, pipeline.inventory.snapshotStock()[Sku("SKU-1")])
    }

    @Test
    fun insufficientStockFailsWithoutChargeOrPersist() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 1), decline = false)

        val result = pipeline.useCase.execute(listOf(line("SKU-1", 5, 500)), "idem-2")

        val failure = assertInstanceOf(PlaceOrderResult.Failure::class.java, result)
        val shortage = assertInstanceOf(InsufficientStockError::class.java, failure.error)
        assertEquals(1, shortage.available)
        assertTrue(pipeline.payments.chargedOrders().isEmpty())
        assertTrue(pipeline.repository.saved().isEmpty())
        assertEquals(1, pipeline.inventory.snapshotStock()[Sku("SKU-1")])
    }

    @Test
    fun paymentDeclineReleasesReservationWithoutPersist() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = true)

        val result = pipeline.useCase.execute(listOf(line("SKU-1", 1, 500)), "idem-3")

        val failure = assertInstanceOf(PlaceOrderResult.Failure::class.java, result)
        val declined = assertInstanceOf(PaymentDeclinedError::class.java, failure.error)
        assertTrue(declined.reason.contains("declined"))
        assertEquals(1, pipeline.payments.chargedOrders().size)
        assertTrue(pipeline.repository.saved().isEmpty())
        assertEquals(10, pipeline.inventory.snapshotStock()[Sku("SKU-1")])
    }

    @Test
    fun saveFailureRefundsAndReleases() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = false)
        pipeline.repository.failSave(true)

        val result = pipeline.useCase.execute(listOf(line("SKU-1", 1, 500)), "idem-4")

        val failure = assertInstanceOf(PlaceOrderResult.Failure::class.java, result)
        assertInstanceOf(PersistenceConflictError::class.java, failure.error)
        assertEquals(1, pipeline.payments.refunded().size)
        assertEquals(10, pipeline.inventory.snapshotStock()[Sku("SKU-1")])
        assertTrue(pipeline.repository.saved().isEmpty())
    }

    @Test
    fun compensationFailureAfterSaveFailure() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = false)
        pipeline.repository.failSave(true)
        pipeline.payments.failRefund(true)

        val result = pipeline.useCase.execute(listOf(line("SKU-1", 1, 500)), "idem-5")

        val failure = assertInstanceOf(PlaceOrderResult.Failure::class.java, result)
        assertInstanceOf(CompensationFailureError::class.java, failure.error)
    }

    @Test
    fun releaseFailureAfterSaveFailureIsCompensationFailure() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = false)
        pipeline.repository.failSave(true)
        pipeline.inventory.failRelease(true)

        val result = pipeline.useCase.execute(listOf(line("SKU-1", 1, 500)), "idem-5c")

        val failure = assertInstanceOf(PlaceOrderResult.Failure::class.java, result)
        assertInstanceOf(CompensationFailureError::class.java, failure.error)
        assertEquals(1, pipeline.payments.refunded().size)
    }

    @Test
    fun releaseFailureAfterDeclineIsCompensationFailure() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = true)
        pipeline.inventory.failRelease(true)

        val result = pipeline.useCase.execute(listOf(line("SKU-1", 1, 500)), "idem-5b")

        val failure = assertInstanceOf(PlaceOrderResult.Failure::class.java, result)
        assertInstanceOf(CompensationFailureError::class.java, failure.error)
    }

    @Test
    fun invalidLinesFailValidationWithoutTouchingPorts() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = false)

        val result = pipeline.useCase.execute(emptyList(), "idem-6")

        val failure = assertInstanceOf(PlaceOrderResult.Failure::class.java, result)
        assertInstanceOf(InvalidOrderError::class.java, failure.error)
        assertTrue(pipeline.payments.chargedOrders().isEmpty())
        assertTrue(pipeline.repository.saved().isEmpty())
        assertEquals(10, pipeline.inventory.snapshotStock()[Sku("SKU-1")])
    }

    @Test
    fun getReturnsNullForUnknownIdentifier() {
        val repository = InMemoryOrderRepository()
        assertNull(repository.get(OrderId("missing-order")))
    }

    @Test
    fun reserveAllIsIdempotentForTheSameKey() {
        val inventory = InMemoryInventoryGateway(mapOf(Sku("SKU-1") to 2))
        val lines = listOf(line("SKU-1", 1, 100))
        val first = inventory.reserveAll(OrderId("ord-1"), lines, "idem-x")
        val second = inventory.reserveAll(OrderId("ord-1"), lines, "idem-x")
        assertInstanceOf(ReservationOutcome.Reserved::class.java, first)
        assertInstanceOf(ReservationOutcome.Reserved::class.java, second)
        assertEquals(1, inventory.snapshotStock()[Sku("SKU-1")])
    }

    @Test
    fun releaseOfUnknownTokenSucceeds() {
        val inventory = InMemoryInventoryGateway(emptyMap())
        val outcome = inventory.release(ReservationToken(OrderId("ord-1"), "missing"))
        assertInstanceOf(ReleaseOutcome.Released::class.java, outcome)
    }

    @Test
    fun chargeIsIdempotentForTheSameKey() {
        val payments = FakePaymentProcessor(decline = false)
        val order = Order(OrderId("ord-1"), listOf(line("SKU-1", 1, 100)))
        val first = payments.charge(order, "idem-pay")
        val second = payments.charge(order, "idem-pay")
        assertInstanceOf(ChargeOutcome.Charged::class.java, first)
        assertInstanceOf(ChargeOutcome.Charged::class.java, second)
        assertEquals(1, payments.chargedOrders().size)
    }

    @Test
    fun saveThenGetReturnsDetachedPaidSnapshot() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = false)
        val result = pipeline.useCase.execute(listOf(line("SKU-1", 1, 100)), "idem-repo")
        val success = result as PlaceOrderResult.Success

        val loaded = pipeline.repository.get(success.order.id)!!
        loaded.ship()
        val stored = pipeline.repository.get(success.order.id)!!
        assertEquals(Order.Status.PAID, stored.status)
        assertEquals(1, stored.version)
    }

    @Test
    fun saveWithStaleVersionIsPersistenceConflict() {
        val repository = InMemoryOrderRepository()
        val order = Order(OrderId("ord-2"), listOf(line("SKU-1", 1, 100)))
        order.pay()
        assertInstanceOf(SaveOutcome.Saved::class.java, repository.save(order, 0))
        val retry = order.snapshot()
        retry.ship()
        assertInstanceOf(SaveOutcome.Conflict::class.java, repository.save(retry, 0))
        assertInstanceOf(SaveOutcome.Saved::class.java, repository.save(order, 1))
    }

    @Test
    fun idempotentReplayDoesNotDoubleCharge() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = false)
        val first = pipeline.useCase.execute(listOf(line("SKU-1", 2, 300)), "idem-7")
        val second = pipeline.useCase.execute(listOf(line("SKU-1", 2, 300)), "idem-7")

        assertInstanceOf(PlaceOrderResult.Success::class.java, first)
        assertInstanceOf(PlaceOrderResult.Success::class.java, second)
        assertEquals(Order.Status.PAID, (second as PlaceOrderResult.Success).order.status)
        assertEquals(1, pipeline.payments.chargedOrders().size)
        assertEquals(8, pipeline.inventory.snapshotStock()[Sku("SKU-1")])
    }

    @Test
    fun reusedKeyWithDifferentPayloadIsInvalidOrder() {
        val pipeline = Pipeline.create(mapOf("SKU-1" to 10), decline = false)
        pipeline.useCase.execute(listOf(line("SKU-1", 1, 100)), "idem-8")

        val result = pipeline.useCase.execute(listOf(line("SKU-1", 2, 100)), "idem-8")

        val failure = assertInstanceOf(PlaceOrderResult.Failure::class.java, result)
        val invalid = assertInstanceOf(InvalidOrderError::class.java, failure.error)
        assertTrue(invalid.reason.contains("idempotency"))
        assertEquals(1, pipeline.payments.chargedOrders().size)
    }

    @Test
    fun concurrentReservationsDoNotOversell() {
        val pipeline = Pipeline.create(mapOf("SKU-A" to 5), decline = false)
        val lines = listOf(line("SKU-A", 5, 100))
        val tasks =
            listOf<Callable<PlaceOrderResult>>(
                Callable { pipeline.useCase.execute(lines, "idem-9a") },
                Callable { pipeline.useCase.execute(lines, "idem-9b") }
            )

        val executor = Executors.newFixedThreadPool(2)
        try {
            val results = executor.invokeAll(tasks).map { join(it) }
            val successes = results.count { it is PlaceOrderResult.Success }
            val shortages =
                results.count { result ->
                    result is PlaceOrderResult.Failure && result.error is InsufficientStockError
                }
            assertEquals(1, successes)
            assertEquals(1, shortages)
        } finally {
            executor.shutdownNow()
        }
        assertEquals(0, pipeline.inventory.snapshotStock()[Sku("SKU-A")])
        assertEquals(1, pipeline.payments.chargedOrders().size)
    }

    private data class Pipeline(
        val useCase: PlaceOrderUseCase,
        val inventory: InMemoryInventoryGateway,
        val payments: FakePaymentProcessor,
        val repository: InMemoryOrderRepository
    ) {
        companion object {
            fun create(
                stock: Map<String, Int>,
                decline: Boolean
            ): Pipeline {
                val initialStock = stock.mapKeys { Sku(it.key) }
                val inventory = InMemoryInventoryGateway(initialStock)
                val payments = FakePaymentProcessor(decline)
                val repository = InMemoryOrderRepository()
                val seq = AtomicInteger()
                val ids = OrderIdGenerator { OrderId("ord-${seq.incrementAndGet()}") }
                return Pipeline(
                    PlaceOrderUseCase(inventory, payments, repository, ids),
                    inventory,
                    payments,
                    repository
                )
            }
        }
    }

    companion object {
        private const val USD = "USD"

        private fun line(
            sku: String,
            quantity: Int,
            minorUnits: Long
        ): OrderLine = OrderLine(Sku(sku), Quantity(quantity), Money(minorUnits, USD))

        private fun join(future: Future<PlaceOrderResult>): PlaceOrderResult =
            try {
                future.get()
            } catch (error: InterruptedException) {
                Thread.currentThread().interrupt()
                throw AssertionError(error)
            } catch (error: java.util.concurrent.ExecutionException) {
                throw AssertionError(error)
            }
    }
}
